From 0967f6b5d6e121680bf1eaa1eda6bcf656bf2e30 Mon Sep 17 00:00:00 2001
From: Karthikeyan K <karthikk@habana.ai>
Date: Fri, 04 Jun 2021 08:17:58 +0300
Subject: [PATCH] [SW-33054] greco: erratum: RR to block  pcie msi

Access to pcie msi region is to be blocked as a workaround for H5 errata
GOYA2_0100.

This patch implements the range register for the same.

Change-Id: I4846b0bff37e487b6a57ac02a0460647618bb8ed
Signed-off-by: Karthikeyan K <karthikk@habana.ai>
---

diff --git a/arch/greco/security.c b/arch/greco/security.c
index ca5a5e8..9ee0be3 100644
--- a/arch/greco/security.c
+++ b/arch/greco/security.c
@@ -65,6 +65,7 @@
 };
 
 enum greco_sec_if_lbw_rrs {
+	LBW_RR_PCIE_MSI,
 	LBW_MAX_RR_IDX = 18,
 };
 
@@ -281,7 +282,7 @@
 	preboot_secure_lbw_rr_wrap(rr_idx, min, max, false);
 }
 
-__unused static void preboot_secure_lbw_rr_short(int rr_idx, u64 min64, u64 max64)
+static void preboot_secure_lbw_rr_short(int rr_idx, u64 min64, u64 max64)
 {
 	u32 min = (min64 & ~HL_CFG_BASE) >> SHIFT_4K_POS;
 	u32 max = (max64 & ~HL_CFG_BASE) >> SHIFT_4K_POS;
@@ -362,6 +363,7 @@
 
 static void preboot_secure_lbw(void)
 {
+	u64 min, max;
 	/*
 	 * There are 18(0..17 rr idx) RRs for LBW.
 	 * The classification and apis are as listed below:
@@ -376,6 +378,14 @@
 	 * NOTE: The apis listed above will block all read and write
 	 * transactions for the input address range.
 	 */
+
+	/*
+	 * Workaround for H5 Errata: GOYA2_0100
+	 * Block PCIE_MSI : RR(LBW_RR_PCIE_MSI)
+	 */
+	min = mmPCIE_MSI_BASE;
+	max = mmPCIE_MSI_BASE + PCIE_MSI_SECTION;
+	preboot_secure_lbw_rr_short(LBW_RR_PCIE_MSI, min, max);
 }
 
 static void preboot_secure_hbw(void)
